src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/FilterResult.cs

```diff
@@ -21,42 +21,18 @@ public FilterResult(CompletionItem completionItem, string filterText, bool match
         }
 
         public int CompareTo(FilterResult other)
-        {
-            var item1 = this.CompletionItem;
-            var item2 = other.CompletionItem;
-
-            var prefixLength1 = item1.FilterText.GetCaseInsensitivePrefixLength(this.FilterText);
-            var prefixLength2 = item2.FilterText.GetCaseInsensitivePrefixLength(other.FilterText);
-
-            // Prefer the item that matches a longer prefix of the filter text.
-            if (prefixLength1 != prefixLength2)
-            {
-                return prefixLength1.CompareTo(prefixLength2);
-            }
-
-            // If the lengths are the same, prefer the one with the higher match priority.
-            // But only if it's an item that would have been hard selected.  We don't want
-            // to aggressively select an item that was only going to be softly offered.
-            var item1Priority = item1.Rules.SelectionBehavior == CompletionItemSelectionBehavior.HardSelection
-                ? item1.Rules.MatchPriority : MatchPriority.Default;
-            var item2Priority = item2.Rules.SelectionBehavior == CompletionItemSelectionBehavior.HardSelection
-                ? item2.Rules.MatchPriority : MatchPriority.Default;
+            => IComparableHelper.CompareTo(this, other, GetComparisonComponents);
 
-            if (item1Priority != item2Priority)
-            {
-                return item1Priority.CompareTo(item2Priority);
-            }
-
-            prefixLength1 = item1.FilterText.GetCaseSensitivePrefixLength(this.FilterText);
-            prefixLength2 = item2.FilterText.GetCaseSensitivePrefixLength(other.FilterText);
-
-            // If there are "Abc" vs "abc", we should prefer the case typed by user.
-            if (prefixLength1 != prefixLength2)
-            {
-                return prefixLength1.CompareTo(prefixLength2);
-            }
-
-            return this.CompletionItem.IsPreferredItem().CompareTo(other.CompletionItem.IsPreferredItem());
+        private static IEnumerable<IComparable> GetComparisonComponents(FilterResult filterResult)
+        {
+            var completionItem = filterResult.CompletionItem;
+            yield return completionItem.FilterText.GetCaseInsensitivePrefixLength(filterResult.FilterText);
+            yield return completionItem.Rules.SelectionBehavior == CompletionItemSelectionBehavior.HardSelection
+                ? filterResult.CompletionItem.Rules.MatchPriority
+                : MatchPriority.Default;
+
+            yield return completionItem.FilterText.GetCaseSensitivePrefixLength(filterResult.FilterText);
+            yield return completionItem.IsPreferredItem();
```

Comment:
definitely concerned.  i think allocations in these paths are not ok.  I think you should have to overloads.  One that takes just an `ImmutableArray<IComparable>`s.  Another that takes an `ImmutableArray<IComparableWithState>` and a `state` arg.  That way this can be allocated once, and you can call the helper from here passing along `filterResult` as an arg that can feed through.  

Does that make sense?

Reply:
Using immutable arrays, we exclude algorithms with variable length for comparison paths like https://github.com/dotnet/roslyn/pull/37894/files#diff-322fd889d857cd92d680c3f841a26094R43

Reply:
What is `IComparableWithState`?

Reply:
somehting you would have to define, but it would allow you to avoid captures by effectively stating that the delegate can be passed extra data.  that way you can statically create the array of delegates, but pass data it needs dynamically.

Reply:
Thank you, @CyrusNajmabadi ! Hopefully implemented `ComparerWithState`. 

@CyrusNajmabadi  and @genlu  please take a look.